src/runtime/policy/ingress-validation.ts

```diff
@@ -14,37 +14,179 @@ import {Capability, Capabilities} from '../capabilities.js';
+
+  constructor(public readonly policies: Policy[]) {
+    for (const policy of this.policies) {
+      for (const target of policy.targets) {
+        const capabilities = target.toCapabilities();
+        for (const field of target.fields) {
+          this.initCapabilitiesMap([target.schemaName], field, capabilities);
+        }
+      }
+    }
+  }
+
+  private initCapabilitiesMap(fieldPaths: string[], field: PolicyField, capabilities: Capabilities[]) {
+    const paths = [...fieldPaths, field.name];
+    if (field.subfields.length === 0) {
+      this.addCapabilities(paths.join('.'), capabilities);
+    }
+    for (const subField of field.subfields) {
+      this.initCapabilitiesMap(paths, subField, capabilities);
+    }
+  }
+
+  private addCapabilities(fieldPath: string, capabilities: Capabilities[]) {
+    if (!this.capabilityByFieldPath.has(fieldPath)) {
+      this.capabilityByFieldPath.set(fieldPath, []);
+    }
+    this.capabilityByFieldPath.get(fieldPath).push(...capabilities);
+  }
+
+  getFieldCapabilities(fieldPath: string): Capabilities[]|undefined {
+    return this.capabilityByFieldPath.get(fieldPath);
+  }
+
+  // Returns success, if all `create` handles in the recipe are declared with
+  // capabilities that comply with the set of policies. Otherwise, returns a
+  // combination of all encountered errors.
+  validateIngressCapabilities(recipe: Recipe): IngressValidationResult {
+    return IngressValidationResult.every(recipe,
+        recipe.handles.filter(h => h.fate === 'create')
+            .map(h => this.validateHandleCapabilities(h)));
+  }
+
+  // Returns success, if the type of the handle, restricted according with the
+  // set of policies, has capabilities that are compliant with the policies'
+  // retention and maxAge permissions. Otherwise, returns a combination of all
+  // encountered errors.
+  private validateHandleCapabilities(handle: Handle): IngressValidationResult {
+    assert(handle.type.maybeEnsureResolved());
+    const result = IngressValidationResult.success(handle);
+    const restrictedType = this.restrictType(handle.type.resolvedType());
+    if (restrictedType) {
+      const fieldPaths = [];
+      for (const [fieldName, field] of Object.entries(restrictedType.getEntitySchema().fields)) {
+        this.collectFieldPaths(restrictedType.getEntitySchema().name, fieldPaths, fieldName, field);
+      }
+
+      for (const fieldPath of fieldPaths) {
+        const fieldCapabilities = this.getFieldCapabilities(fieldPath);
+        assert(fieldCapabilities, `Missing capabilities for ${fieldPath}`);
+        const fieldResults = fieldCapabilities.map(
+            fc => fc.isAllowedForIngress(handle.capabilities));
+        if (!fieldResults.some(r => r.success)) {
+          result.addResult(IngressValidationResult.failWith(handle,
+              `Failed validating ingress for field '${fieldPath}'` +
+              ` of '${handle.id ? handle.id : handle.connections[0].getQualifiedName()}`,
```

Comment:
Could be: `handle.id || handle.connections......`

Reply:
done